File: UnitySokoban/Assets/Scripts/Planning/IterativeWidthPlanner/IteratedWidthSearch.cs
using StateSpaceSearchProject;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Planning;

namespace IterativeWidthPlanner
{
    public class IteratedWidthSearch : StateSpaceSearch
    {
        public readonly int maxWidth;
        private readonly Queue<StateSpaceNode> queue = new Queue<StateSpaceNode>();

        public IteratedWidthSearch(StateSpaceProblem problem, int maxWidth)
            :base(problem)
        {
            this.maxWidth = maxWidth;
            queue.Enqueue(root);
        }

        public override Plan findNextSolution()
        {
            while (queue.Count != 0)
            {
                StateSpaceNode node = queue.Dequeue();
                node.expand();
                foreach (StateSpaceNode child in node.children)
                    if (!prune(child))
                        queue.Enqueue(child);
                if (problem.goal.IsTrue(node.state))
                    return node.plan;
            }
            return null;
        }

        private bool prune(StateSpaceNode node) {
            for (int i = 1; i <= maxWidth; i++)
                if (Novelty.hasNovelty((StateSpaceProblem)problem, node, i))
                    return false;
            return true;
        }
    }
}

File: Iterative Width Planner (sgware)/src/edu/uno/ai/planning/iw/IteratedWidthSearch.java
package edu.uno.ai.planning.iw;

import java.util.LinkedList;
import java.util.Queue;

import edu.uno.ai.planning.Plan;
import edu.uno.ai.planning.ss.StateSpaceNode;
import edu.uno.ai.planning.ss.StateSpaceProblem;
import edu.uno.ai.planning.ss.StateSpaceSearch;

public class IteratedWidthSearch extends StateSpaceSearch {

	public final int maxWidth;
	private final Queue<StateSpaceNode> queue = new LinkedList<>();
	
	public IteratedWidthSearch(StateSpaceProblem problem, int maxWidth) {
		super(problem);
		this.maxWidth = maxWidth;
		queue.add(root);
	}

	@Override
	public Plan findNextSolution() {
		while(!queue.isEmpty()) {
			StateSpaceNode node = queue.poll();
			node.expand();
			for(StateSpaceNode child : node.children)
				if(!prune(child))
					queue.add(child);
			if(problem.goal.isTrue(node.state))
				return node.plan;
		}
		return null;
	}
	
	private final boolean prune(StateSpaceNode node) {
		for(int i=1; i<=maxWidth; i++)
			if(Novelty.hasNovelty(problem, node, i))
				return false;
		return true;
	}
}